src/lists.rs

```diff
@@ -130,24 +159,50 @@ pub fn write_list<'b>(items: &[(String, String)], formatting: &ListFormatting<'b
             _ => {}
         }
 
-        result.push_str(item);
+        // Pre-comments
+        if let Some(ref comment) = item.pre_comment {
+            result.push_str(&rewrite_comment(comment,
+                                             false,
+                                             1000,
+                                             formatting.indent));
+            result.push('\n');
+            result.push_str(indent_str);
+        }
 
-        if tactic != ListTactic::Vertical && comment.len() > 0 {
-            if !comment.starts_with('\n') {
-                result.push(' ');
-            }
-            result.push_str(comment);
+        result.push_str(&item.item);
+
+        // Post-comments
+        if tactic != ListTactic::Vertical && item.post_comment.is_some() {
+            // We'll assume it'll fit on one line at this point
+            let formatted_comment = rewrite_comment(item.post_comment.as_ref().unwrap(),
+                                                    true,
+                                                    1000,
+                                                    0);
+
+            result.push(' ');
+            result.push_str(&formatted_comment);
         }
 
         if separate {
             result.push_str(formatting.separator);
         }
 
-        if tactic == ListTactic::Vertical && comment.len() > 0 {
-            if !comment.starts_with('\n') {
-                result.push(' ');
-            }
-            result.push_str(comment);
+        if tactic == ListTactic::Vertical && item.post_comment.is_some() {
+            let width = formatting.v_width - item_width - formatting.separator.len();
+            let offset = formatting.indent + item_width + formatting.separator.len();
+            let comment = item.post_comment.as_ref().unwrap();
+            // Use block-style only for the last item or multiline comments
```

Comment:
I think that in the case of a post-comment, where the user has used /* */ and it is a multiline comment, then it is OK to leave it as a /* */ comment and don't convert to //.

I think we should strongly avoid turning post-comments into pre-comments, I'd rather err on the side of sub-optimal formatting than do that.

I'm also a bit wary about moving post-comments to pre-comments. We should perhaps do this at some point, but I'm wary about diving right in - when we do formatting which is more than just changing whitespace, I think there is a higher bar for 'correctness'.

I also like the invariant (which had not previously occurred to me) that the output of rustfmt should be idempotent under further reformatting.